# THYROID DISEASE DATASET PREPARATION & CLEANING

In [52]:
# Import libraries that will assist in cleaning

# Pandas for data manipulation and numpy for numeric operations

import pandas as pd
import numpy as np

In [53]:
# Load the thyroid dataset

data1 = pd.read_csv('/Users/RyanMburu/Desktop/DS Projects/Thyroid Disease Detector/hypothyroid.csv')

# Print out first 5 rows
data1.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0  41   F            f                  f                         f    f   
1  23   F            f                  f                         f    f   
2  46   M            f                  f                         f    f   
3  70   F            t                  f                         f    f   
4  70   F            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid  ... TT4 measured  \
0        f               f              f                 f  ...            t   
1        f               f              f                 f  ...            t   
2        f               f              f                 f  ...            t   
3        f               f              f                 f  ...            t   
4        f               f              f                 f  ...            t   

   TT4 T4U measured   T4U FTI measured  FTI TBG measured TBG referral source  \
0  125            t  1.14            t  109            f   ?            SVHC   
1  102            f     ?            f    ?            f   ?           other   
2  109            t  0.91            t  120            f   ?           other   
3  175            f     ?            f    ?            f   ?           other   
4   61            t  0.87            t   70            f   ?             SVI   

  binaryClass  
0           P  
1           P  
2           P  
3           P  
4           P  

[5 rows x 30 columns]

In [54]:
# Print out last 5 rows
data1.tail()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
3767  30   F            f                  f                         f    f   
3768  68   F            f                  f                         f    f   
3769  74   F            f                  f                         f    f   
3770  72   M            f                  f                         f    f   
3771  64   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  ...  \
3767        f               f              f                 f  ...   
3768        f               f              f                 f  ...   
3769        f               f              f                 f  ...   
3770        f               f              f                 f  ...   
3771        f               f              f                 f  ...   

     TT4 measured  TT4 T4U measured   T4U FTI measured  FTI TBG measured TBG  \
3767            f    ?            f     ?            f    ?            f   ?   
3768            t  124            t  1.08            t  114            f   ?   
3769            t  112            t  1.07            t  105            f   ?   
3770            t   82            t  0.94            t   87            f   ?   
3771            t   99            t  1.07            t   92            f   ?   

     referral source binaryClass  
3767           other           P  
3768             SVI           P  
3769           other           P  
3770             SVI           P  
3771           other           P  

[5 rows x 30 columns]

In [55]:
# Gather number of rows and columns present

data1.shape

(3772, 30)

The Hypothyroid dataset has 3,770 rows of data with 30 columns

In [56]:
# Check datatypes
data1.dtypes

age                          object
sex                          object
on thyroxine                 object
query on thyroxine           object
on antithyroid medication    object
sick                         object
pregnant                     object
thyroid surgery              object
I131 treatment               object
query hypothyroid            object
query hyperthyroid           object
lithium                      object
goitre                       object
tumor                        object
hypopituitary                object
psych                        object
TSH measured                 object
TSH                          object
T3 measured                  object
T3                           object
TT4 measured                 object
TT4                          object
T4U measured                 object
T4U                          object
FTI measured                 object
FTI                          object
TBG measured                 object
TBG                         

# DATA CLEANING

The data cleaning stage will be separated into 5 stages

1. Validity
2. Accuracy
3. Consistency
4. Completeness
5. Uniformity

## 1. Validity

In [57]:
# Will remove columns useless for our analysis and building ML model

# Columns that say whether TT4, TSH columns are measured or not are redundant as patients with no data do not help our analysis

data1 = data1.drop(columns=['TSH measured', 'T3 measured', 'TT4 measured', 'T4U measured', 'FTI measured', 'TBG measured'], axis=1)
data1.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0  41   F            f                  f                         f    f   
1  23   F            f                  f                         f    f   
2  46   M            f                  f                         f    f   
3  70   F            t                  f                         f    f   
4  70   F            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid  ...  \
0        f               f              f                 f  ...   
1        f               f              f                 f  ...   
2        f               f              f                 f  ...   
3        f               f              f                 f  ...   
4        f               f              f                 f  ...   

  hypopituitary psych   TSH   T3  TT4   T4U  FTI TBG referral source  \
0             f     f   1.3  2.5  125  1.14  109   ?            SVHC   
1             f     f   4.1    2  102     ?    ?   ?           other   
2             f     f  0.98    ?  109  0.91  120   ?           other   
3             f     f  0.16  1.9  175     ?    ?   ?           other   
4             f     f  0.72  1.2   61  0.87   70   ?             SVI   

  binaryClass  
0           P  
1           P  
2           P  
3           P  
4           P  

[5 rows x 24 columns]

In [58]:
# Will drop TBG column as well as its just question marks
data1 = data1.drop(columns=['TBG'], axis=1)
data1.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0  41   F            f                  f                         f    f   
1  23   F            f                  f                         f    f   
2  46   M            f                  f                         f    f   
3  70   F            t                  f                         f    f   
4  70   F            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     f   
4        f               f              f                 f  ...     f   

  hypopituitary psych   TSH   T3  TT4   T4U  FTI referral source binaryClass  
0             f     f   1.3  2.5  125  1.14  109            SVHC           P  
1             f     f   4.1    2  102     ?    ?           other           P  
2             f     f  0.98    ?  109  0.91  120           other           P  
3             f     f  0.16  1.9  175     ?    ?           other           P  
4             f     f  0.72  1.2   61  0.87   70             SVI           P  

[5 rows x 23 columns]

The data is deemed valid.

## 2. Accuracy

- This is where the data is checked for mathematical accuracy.

- Due to my lack of expertise in the medical field, it will be assumed that the data is mathematically accurate and the nymerical data is in the right format.

## 3. Completeness

- This is where the data is ensured to be complete.

- Null Values will be handled here

In [59]:
# Handling null values

data1.isna().sum()

age                          0
sex                          0
on thyroxine                 0
query on thyroxine           0
on antithyroid medication    0
sick                         0
pregnant                     0
thyroid surgery              0
I131 treatment               0
query hypothyroid            0
query hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH                          0
T3                           0
TT4                          0
T4U                          0
FTI                          0
referral source              0
binaryClass                  0
dtype: int64

The pandas operation claims that there are no null values present in the data. This is so as null values in our data are showen by the '?' mark.

- '?' marks will be replaced by NULL

In [60]:
# Replace '?' with NULL

data1 = data1.replace('?', np.NaN)
data1

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0     41   F            f                  f                         f    f   
1     23   F            f                  f                         f    f   
2     46   M            f                  f                         f    f   
3     70   F            t                  f                         f    f   
4     70   F            f                  f                         f    f   
...   ..  ..          ...                ...                       ...  ...   
3767  30   F            f                  f                         f    f   
3768  68   F            f                  f                         f    f   
3769  74   F            f                  f                         f    f   
3770  72   M            f                  f                         f    f   
3771  64   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  ... tumor  \
0           f               f              f                 f  ...     f   
1           f               f              f                 f  ...     f   
2           f               f              f                 f  ...     f   
3           f               f              f                 f  ...     f   
4           f               f              f                 f  ...     f   
...       ...             ...            ...               ...  ...   ...   
3767        f               f              f                 f  ...     t   
3768        f               f              f                 f  ...     f   
3769        f               f              f                 f  ...     f   
3770        f               f              f                 f  ...     f   
3771        f               f              f                 f  ...     f   

     hypopituitary psych   TSH   T3  TT4   T4U  FTI referral source  \
0                f     f   1.3  2.5  125  1.14  109            SVHC   
1                f     f   4.1    2  102   NaN  NaN           other   
2                f     f  0.98  NaN  109  0.91  120           other   
3                f     f  0.16  1.9  175   NaN  NaN           other   
4                f     f  0.72  1.2   61  0.87   70             SVI   
...            ...   ...   ...  ...  ...   ...  ...             ...   
3767             f     f   NaN  NaN  NaN   NaN  NaN           other   
3768             f     f     1  2.1  124  1.08  114             SVI   
3769             f     f   5.1  1.8  112  1.07  105           other   
3770             f     f   0.7    2   82  0.94   87             SVI   
3771             f     f     1  2.2   99  1.07   92           other   

     binaryClass  
0              P  
1              P  
2              P  
3              P  
4              P  
...          ...  
3767           P  
3768           P  
3769           P  
3770           P  
3771           P  

[3772 rows x 23 columns]

In [61]:
# Check sum of null values per column
data1.isna().sum()

age                            1
sex                          150
on thyroxine                   0
query on thyroxine             0
on antithyroid medication      0
sick                           0
pregnant                       0
thyroid surgery                0
I131 treatment                 0
query hypothyroid              0
query hyperthyroid             0
lithium                        0
goitre                         0
tumor                          0
hypopituitary                  0
psych                          0
TSH                          369
T3                           769
TT4                          231
T4U                          387
FTI                          385
referral source                0
binaryClass                    0
dtype: int64

Will replace null values of gender with 'Not Specified' and the rest will be dropped

In [62]:
# Replace null values in sex with 'NS' (Not Specified)

data1['sex'] = data1['sex'].fillna('NS')

In [92]:
data1['binaryClass'].value_counts()

P    3481
N     291
Name: binaryClass, dtype: int64

In [77]:
# Will proceed to drop rows with null values, to avoid inconsistency when modelling

data2 = data1.dropna().reset_index()
data2

index age sex on thyroxine query on thyroxine on antithyroid medication  \
0         0  41   F            f                  f                         f   
1         4  70   F            f                  f                         f   
2         7  80   F            f                  f                         f   
3         8  66   F            f                  f                         f   
4         9  68   M            f                  f                         f   
...     ...  ..  ..          ...                ...                       ...   
2746   3766  19   F            f                  f                         f   
2747   3768  68   F            f                  f                         f   
2748   3769  74   F            f                  f                         f   
2749   3770  72   M            f                  f                         f   
2750   3771  64   F            f                  f                         f   

     sick pregnant thyroid surgery I131 treatment  ... tumor hypopituitary  \
0       f        f               f              f  ...     f             f   
1       f        f               f              f  ...     f             f   
2       f        f               f              f  ...     f             f   
3       f        f               f              f  ...     t             f   
4       f        f               f              f  ...     f             f   
...   ...      ...             ...            ...  ...   ...           ...   
2746    f        f               f              f  ...     f             f   
2747    f        f               f              f  ...     f             f   
2748    f        f               f              f  ...     f             f   
2749    f        f               f              f  ...     f             f   
2750    f        f               f              f  ...     f             f   

     psych   TSH   T3  TT4   T4U  FTI referral source binaryClass  
0        f   1.3  2.5  125  1.14  109            SVHC           P  
1        f  0.72  1.2   61  0.87   70             SVI           P  
2        f   2.2  0.6   80   0.7  115             SVI           P  
3        f   0.6  2.2  123  0.93  132             SVI           P  
4        f   2.4  1.6   83  0.89   93             SVI           P  
...    ...   ...  ...  ...   ...  ...             ...         ...  
2746     f   8.8  2.7  108  1.11   97           other           N  
2747     f     1  2.1  124  1.08  114             SVI           P  
2748     f   5.1  1.8  112  1.07  105           other           P  
2749     f   0.7    2   82  0.94   87             SVI           P  
2750     f     1  2.2   99  1.07   92           other           P  

[2751 rows x 24 columns]

In [78]:
# Drop index column
data2.drop(columns=['index'], axis=1, inplace=True)

In [79]:
# Check null values
data2.isna().sum()

age                          0
sex                          0
on thyroxine                 0
query on thyroxine           0
on antithyroid medication    0
sick                         0
pregnant                     0
thyroid surgery              0
I131 treatment               0
query hypothyroid            0
query hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH                          0
T3                           0
TT4                          0
T4U                          0
FTI                          0
referral source              0
binaryClass                  0
dtype: int64

- About 900 records of data were lost, but our data now has no null values.

- Saved about 200 records by changing null values in gender with NS, as I predict the sex column will not be crucial in predicting the model's output

## 4. Consistency

- This is where duplicate values are inspected and dropped if any.

In [80]:
# Checking for duplicates

data2.duplicated().sum()

1

Only one duplicated value, which will be dropped from our data

In [81]:
# Drop the duplicate
data2 = data2.drop_duplicates()

# Check again if there is any duplicate value
data2.duplicated().sum()

0

No more duplicates. Our data is consistent

## 5. Uniformity

This is where the data should be the same i.e data types and column names are in the same format etc

In [82]:
data2.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0  41   F            f                  f                         f    f   
1  70   F            f                  f                         f    f   
2  80   F            f                  f                         f    f   
3  66   F            f                  f                         f    f   
4  68   M            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     t   
4        f               f              f                 f  ...     f   

  hypopituitary psych   TSH   T3  TT4   T4U  FTI referral source binaryClass  
0             f     f   1.3  2.5  125  1.14  109            SVHC           P  
1             f     f  0.72  1.2   61  0.87   70             SVI           P  
2             f     f   2.2  0.6   80   0.7  115             SVI           P  
3             f     f   0.6  2.2  123  0.93  132             SVI           P  
4             f     f   2.4  1.6   83  0.89   93             SVI           P  

[5 rows x 23 columns]

In [83]:
# Will change last column name to outcome in lowercase for uniformity purposes

data2 = data2.rename(columns={'binaryClass' : 'output'})
data2.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0  41   F            f                  f                         f    f   
1  70   F            f                  f                         f    f   
2  80   F            f                  f                         f    f   
3  66   F            f                  f                         f    f   
4  68   M            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     t   
4        f               f              f                 f  ...     f   

  hypopituitary psych   TSH   T3  TT4   T4U  FTI referral source output  
0             f     f   1.3  2.5  125  1.14  109            SVHC      P  
1             f     f  0.72  1.2   61  0.87   70             SVI      P  
2             f     f   2.2  0.6   80   0.7  115             SVI      P  
3             f     f   0.6  2.2  123  0.93  132             SVI      P  
4             f     f   2.4  1.6   83  0.89   93             SVI      P  

[5 rows x 23 columns]

In [84]:
# Check for datatypes

data2.dtypes

age                          object
sex                          object
on thyroxine                 object
query on thyroxine           object
on antithyroid medication    object
sick                         object
pregnant                     object
thyroid surgery              object
I131 treatment               object
query hypothyroid            object
query hyperthyroid           object
lithium                      object
goitre                       object
tumor                        object
hypopituitary                object
psych                        object
TSH                          object
T3                           object
TT4                          object
T4U                          object
FTI                          object
referral source              object
output                       object
dtype: object

In [85]:
# Will change age, TSH, T3, TT4, T4U and FTI to integer and float datatypes

data2['age'] = data2['age'].astype(int)

data2[['TSH', 'T3', 'TT4', 'T4U', 'FTI']] = data2[['TSH', 'T3', 'TT4', 'T4U', 'FTI']].astype(float)

In [86]:
data2.dtypes

age                            int64
sex                           object
on thyroxine                  object
query on thyroxine            object
on antithyroid medication     object
sick                          object
pregnant                      object
thyroid surgery               object
I131 treatment                object
query hypothyroid             object
query hyperthyroid            object
lithium                       object
goitre                        object
tumor                         object
hypopituitary                 object
psych                         object
TSH                          float64
T3                           float64
TT4                          float64
T4U                          float64
FTI                          float64
referral source               object
output                        object
dtype: object

Data types are uniform and satisfactory.

In [87]:
# Final dataset
data2

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0      41   F            f                  f                         f    f   
1      70   F            f                  f                         f    f   
2      80   F            f                  f                         f    f   
3      66   F            f                  f                         f    f   
4      68   M            f                  f                         f    f   
...   ...  ..          ...                ...                       ...  ...   
2746   19   F            f                  f                         f    f   
2747   68   F            f                  f                         f    f   
2748   74   F            f                  f                         f    f   
2749   72   M            f                  f                         f    f   
2750   64   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  ... tumor  \
0           f               f              f                 f  ...     f   
1           f               f              f                 f  ...     f   
2           f               f              f                 f  ...     f   
3           f               f              f                 f  ...     t   
4           f               f              f                 f  ...     f   
...       ...             ...            ...               ...  ...   ...   
2746        f               f              f                 f  ...     f   
2747        f               f              f                 f  ...     f   
2748        f               f              f                 f  ...     f   
2749        f               f              f                 f  ...     f   
2750        f               f              f                 f  ...     f   

     hypopituitary psych   TSH   T3    TT4   T4U    FTI  referral source  \
0                f     f  1.30  2.5  125.0  1.14  109.0             SVHC   
1                f     f  0.72  1.2   61.0  0.87   70.0              SVI   
2                f     f  2.20  0.6   80.0  0.70  115.0              SVI   
3                f     f  0.60  2.2  123.0  0.93  132.0              SVI   
4                f     f  2.40  1.6   83.0  0.89   93.0              SVI   
...            ...   ...   ...  ...    ...   ...    ...              ...   
2746             f     f  8.80  2.7  108.0  1.11   97.0            other   
2747             f     f  1.00  2.1  124.0  1.08  114.0              SVI   
2748             f     f  5.10  1.8  112.0  1.07  105.0            other   
2749             f     f  0.70  2.0   82.0  0.94   87.0              SVI   
2750             f     f  1.00  2.2   99.0  1.07   92.0            other   

      output  
0          P  
1          P  
2          P  
3          P  
4          P  
...      ...  
2746       N  
2747       P  
2748       P  
2749       P  
2750       P  

[2750 rows x 23 columns]

## Encoding

Will encode the output of the dataset to 1s and 0s, as Machine Learning models take integers only as y input.

In [88]:
# Will encode the output column to 0 and 1

# So as to use as minimal libraries as possible, will not call the hot one encoder library from sklearn

# Will encode traditionally by find and replace method

codes = {'output' : {'P' : 0, 'N' : 1}}

codes

{'output': {'P': 0, 'N': 1}}

In [90]:
# Replace them in dataset

data2= data2.replace(codes)
data2.head()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0   41   F            f                  f                         f    f   
1   70   F            f                  f                         f    f   
2   80   F            f                  f                         f    f   
3   66   F            f                  f                         f    f   
4   68   M            f                  f                         f    f   

  pregnant thyroid surgery I131 treatment query hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     t   
4        f               f              f                 f  ...     f   

  hypopituitary psych   TSH   T3    TT4   T4U    FTI  referral source  output  
0             f     f  1.30  2.5  125.0  1.14  109.0             SVHC       0  
1             f     f  0.72  1.2   61.0  0.87   70.0              SVI       0  
2             f     f  2.20  0.6   80.0  0.70  115.0              SVI       0  
3             f     f  0.60  2.2  123.0  0.93  132.0              SVI       0  
4             f     f  2.40  1.6   83.0  0.89   93.0              SVI       0  

[5 rows x 23 columns]

In [91]:
# Value count the output column
data2['output'].value_counts()

0    2527
1     223
Name: output, dtype: int64

## Exporting dataset

In [ ]:
# Will export clean dataset as a csv file

data2.to_csv('clean_hypothyroid.csv')